packages/parcel-runtime/src/runtimes/background-service-runtime.ts

```diff
@@ -1,80 +1,105 @@
 const parent = module.bundle.parent
 
-if (!parent || !parent.isParcelRequire) {
-  injectHmrSocket(async (updatedAssets) => {
-    const manifestChange = updatedAssets.find((e) => e.type === "json")
-    if (!manifestChange) {
-      return
+const state = {
+  buildReady: false,
+  hmrUpdated: false,
+  csCodeChanged: false,
+  ports: new Set<chrome.runtime.Port>()
+}
+
+function consolidateUpdate(forced = false) {
+  if (
+    forced ||
+    (state.buildReady && (state.hmrUpdated || state.csCodeChanged))
+  ) {
+    vLog("BGSW Runtime - reloading")
+    extCtx.runtime.reload()
+    for (const port of state.ports) {
+      port.postMessage({
+        __plasmo_cs_reload__: true
+      })
     }
+  }
+}
 
-    const changedIdSet = new Set(updatedAssets.map((e) => e.id))
+if (!parent || !parent.isParcelRequire) {
+  const hmrSocket = injectHmrSocket(async (updatedAssets) => {
+    vLog("BGSW Runtime - On HMR Update")
+
+    state.hmrUpdated ||= updatedAssets
+      .filter((asset) => asset.envHash === runtimeData.envHash)
+      .some((asset) => isDependencyOfBundle(module.bundle, asset.id))
+
+    const manifestChange = updatedAssets.find((e) => e.type === "json")
 
-    const deps = Object.values(manifestChange.depsByBundle)
-      .map((o) => Object.values(o))
-      .flat()
+    if (!!manifestChange) {
+      const changedIdSet = new Set(updatedAssets.map((e) => e.id))
 
-    const shouldReload = deps.every((dep) => changedIdSet.has(dep))
+      const deps = Object.values(manifestChange.depsByBundle)
+        .map((o) => Object.values(o))
+        .flat()
 
-    if (shouldReload) {
-      extCtx.runtime.reload()
+      state.hmrUpdated ||= deps.every((dep) => changedIdSet.has(dep))
     }
+
+    consolidateUpdate()
   })
-}
 
-async function runtimeMessageHandler(msg: BackgroundMessage) {
-  if (msg.__plasmo_full_reload__) {
-    extCtx.runtime.reload()
-  }
+  hmrSocket.addEventListener("open", () => {
+    const interval = setInterval(() => hmrSocket.send("ping"), 24_000)
+    hmrSocket.addEventListener("close", () => clearInterval(interval))
+  })
 
-  return true
+  hmrSocket.addEventListener("close", async () => {
+    await pollingDevServer()
+    consolidateUpdate()
+  })
 }
 
-extCtx.runtime.onMessage.addListener(runtimeMessageHandler)
+injectBuilderSocket(async () => {
+  vLog("BGSW Runtime - on build repackaged")
+  // maybe we should wait for a bit until determine if the build is trully ready
+  state.buildReady ||= true
+  consolidateUpdate()
+})
 
 extCtx.runtime.onConnect.addListener(function (port) {
-  if (port.name.startsWith("__plasmo_runtime_")) {
-    port.onMessage.addListener(runtimeMessageHandler)
+  if (port.name.startsWith("__plasmo_runtime_script_")) {
+    state.ports.add(port)
+    port.onDisconnect.addListener(() => {
+      state.ports.delete(port)
+    })
+
+    port.onMessage.addListener(function (msg: BackgroundMessage) {
+      if (msg.__plasmo_cs_changed__) {
+        vLog("BGSW Runtime - on CS code changed")
```

Comment:
same here. maybe you should change the one that has caps since you don't use caps in any of these others